core/store/changeset.go

```diff
@@ -16,3 +26,43 @@ type KVPair struct {
 	// Remove is true when the key must be removed from state.
 	Remove bool
 }
+
+func NewChangeset() *Changeset {
+	return &Changeset{}
+}
+
+func NewChangesetWithPairs(pairs map[string]KVPairs) *Changeset {
+	return &Changeset{}
+}
+
+// Size returns the number of key-value pairs in the batch.
+func (cs *Changeset) Size() int {
+	cnt := 0
+	for _, pairs := range cs.Changes {
+		cnt += len(pairs.StateChanges)
+	}
+
+	return cnt
+}
+
+// Add adds a key-value pair to the ChangeSet.
+func (cs *Changeset) Add(storeKey, key, value []byte, remove bool) {
```

Comment:
the remove semantics are not applied

Reply:
dont follow, they are present

Reply:
previously, assuming value = nil is Remove, any concern of it?

Reply:
>previously, assuming value = nil is Remove, any concern of it?

If this is true, and `KVPair.Value = nil` should carry an implied meaning of delete to SS and SC, then why does the field `KVPair.Remove bool` exist at all? It also means we cannot present a key with an nil value?

Reply:
exactly. I've expressed numerous times `Remove bool` should be removed. It's not necessary.

Reply:
yes, the value cannot be a nil, instead using []byte{}

Reply:
not particularly attached to `Remove`  being present or not, i just think that for a public API is better to be explicit on things than implicit,also value being nil equalling a removal is not consistent with the go language itself, eg: https://go.dev/play/p/XoD_09G8Drw
setting a nil value in a map does not equal to removal. 

but as i said it's fine either way to me

Reply:
Agreed that `key=nil -> delete` is an odd idiom, adjusting is also state breaking in IAVL.  If we're going to change now, server v2 is probably the time to.

Reply:
seems we are good with moving forward with the current api?